Kafka-16355: Fix ConcurrentModificationException in evictWhile Method

abhi-ksolves opened the pull request:
This PR addresses an issue with the evictWhile method where a ConcurrentModificationException was being thrown due to modifications made to sortedMap during iteration. The following changes have been made:

- Safe Iteration: The method now iterates over a copied list of entries from sortedMap, ensuring that modifications to sortedMap do not interfere with the iteration process.
- Consistent Removal: Entries are safely removed from sortedMap during the iteration without causing a ConcurrentModificationException

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

abhi-ksolves committed:
Fix ConcurrentModificationException in evictWhile Method

aliehsaeedii commented:
@abhi-ksolves Thanks for the PR.
I don't see the point where the code is modifying the `sortedMap` while iterating over it. In `evictWhile` method, we are already using the iterators, which is the solution for not getting the `ConcurrentModificationException`. Are you sure that the exception you got, is from this part of the code?

aliehsaeedii commented:
@abhi-ksolves Thanks for the fix. 
For a more optimized implementation, could we create a list of to be deleted `BufferKey`s and  remove them from the `sortedMap` after the loop?

mjsax commented:
@abhi-ksolves -- What is the status of this PR? Are you still interested to finish it?

mjsax commented:
@abhi-ksolves -- Any updates?

abhi-ksolves commented:
Hi @mjsax 
Sorry for the delay; I got involved in a lot of things. 

Will start to review this soon.

abhi-ksolves committed:
Suggested Fixes

abhi-ksolves commented:
Hi @mjsax 
Following up for review.

abhi-ksolves commented:
Comment to keep this PR Active 
Review requested @mjsax

aliehsaeedii commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueChangeBuffer.java:
```diff
@@ -381,17 +383,19 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
     public void evictWhile(final Supplier<Boolean> predicate,
                            final Consumer<Eviction<K, Change<V>>> callback) {
-        final Iterator<Map.Entry<BufferKey, BufferValue>> delegate = sortedMap.entrySet().iterator();
+        final List<Map.Entry<BufferKey, BufferValue>> entries = new ArrayList<>(sortedMap.entrySet());
```
Could we keep the `Iterator`? The list is wasting memory and it was the reason for suggesting a smaller list for the `keysToDelete` rather of copying the whole sortedMap.

aliehsaeedii reviewed (commented):
Thanks, @abhi-ksolves, for following up. I left a comment.

Going back to the ticket, I was wondering why the CME happened. While we are looping through an iterator, even if we modify it, CME should not happen. Should we first find out why it happened? What was accessing sortedMap at the same time? The issue is mostly due to `index.remove(key)`, maybe or sth else?

aliehsaeedii commented:
@abhi-ksolves -- gentle reminder

abhi-ksolves commented:
Thanks for the feedback @aliehsaeedii.

You are right regarding the memory overhead.

Regarding the CME (ConcurrentModificationException):

- Why it happens: CME occurs if we modify sortedMap directly (e.g., sortedMap.remove()) while iterating over it. It does not happen if we use iterator.remove().

- Index map: Modifying the index map (via index.remove()) is safe and does not cause CME in the sortedMap iterator because they are separate collections.

I have updated the method implementation to use the Iterator directly as suggested. I also simplified the minTimestamp update logic to occur once at the end of the loop, rather than recalculating it on every iteration.

PTAL into code snippet

```
   public void evictWhile(final Supplier<Boolean> predicate,
                           final Consumer<Eviction<K, Change<V>>> callback) {
        final Iterator<Map.Entry<BufferKey, BufferValue>> iterator = sortedMap.entrySet().iterator();
        int evictions = 0;

        while (iterator.hasNext()) {
            if (!predicate.get()) {
                break;
            }

            final Map.Entry<BufferKey, BufferValue> next = iterator.next();
            final BufferValue bufferValue = next.getValue();

            final K key = keySerde.deserializer().deserialize(changelogTopic, next.getKey().key().get());
            final Change<V> value = valueSerde.deserializeParts(
                changelogTopic,
                new Change<>(bufferValue.newValue(), bufferValue.oldValue())
            );

            callback.accept(new Eviction<>(key, value, bufferValue.context()));

            if (loggingEnabled) {
                dirtyKeys.add(next.getKey().key());
            }

            memBufferSize -= computeRecordSize(next.getKey().key(), bufferValue);

            iterator.remove();

            index.remove(next.getKey().key());

            evictions++;
        }

        if (evictions > 0) {
            minTimestamp = sortedMap.isEmpty() ? Long.MAX_VALUE : sortedMap.firstKey().time();
            updateBufferMetrics();
        }
    }
```

squah-confluent commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueChangeBuffer.java:
```diff
@@ -381,17 +383,19 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
     public void evictWhile(final Supplier<Boolean> predicate,
                            final Consumer<Eviction<K, Change<V>>> callback) {
-        final Iterator<Map.Entry<BufferKey, BufferValue>> delegate = sortedMap.entrySet().iterator();
+        final List<Map.Entry<BufferKey, BufferValue>> entries = new ArrayList<>(sortedMap.entrySet());
```
We need to understand why the `ConcurrentModificationException` happened. If it's due to another thread updating the collection, then we can still encounter a `ConcurrentModificationException` in the `ArrayList` constructor because it calls `sortedMap.entrySet().toArray()` ... which uses the iterator.

abhi-ksolves replied:
As per my thought, @squah-confluent, this store is used within a Kafka Streams StreamTask, which guarantees single-threaded execution. We do not expect concurrent access from other threads, so the ConcurrentModificationException is not due to a race condition (threading issue).

Any CME encountered here would be strictly due to modifying the underlying collection while iterating over it within the same thread (e.g., if a previous implementation used sortedMap.remove() directly instead of iterator.remove()).

Since we are in a single-threaded context, the ArrayList copy was just an inefficient way to avoid loop modification issues. Switching to the Iterator approach with iterator.remove() as suggested by @aliehsaeedii is seems to be memory-efficient fix that adheres to the single-threaded contract of the StreamTask.

I have added a new method implementation in the same thread above.

aliehsaeedii commented:
@abhi-ksolves Thanks for follwing up. 
As you mentioned what ends up CME is in fact direct manipulation of the `sortedMap` which could be removing or putting. The question was where does that happen? In the former rounds, I missed the `callback`. It seems that `callback` is the culprit. This is the scenario:
1. evictWhile is iterating over the sortedMap                                                                                                                                                             
2. It calls the `emit` callback  (see `evictwhile` call in `KTableSuppressProcessorSupplier`)                                                                                                                       
4. `emit` calls `context.forward()` which pushes the record downstream                                                                                                                                       
5. Downstream processing can trigger additional operations such as remvoing from or putting to sortedMap                                                                                                                                                                                                                                                                                    
6. When control returns to the evictWhile, the iterator detects the modification and throws ConcurrentModificationException  

This is particularly problematic during state restoration (as mentioned in the ticket) because, restoration processes many records in batches.                                                                                                                                                                                                                                                                            

So the solution must put the `callback` outside of the iteration loop.

aliehsaeedii commented:
@abhi-ksolves --gentle reminder:)

abhi-ksolves committed:
Reviewer Corrections

abhi-ksolves commented:
@aliehsaeedii 
Taken another attempt 
Pushed a new commit.

PTAL

aliehsaeedii commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueChangeBuffer.java:
```diff
@@ -381,55 +383,46 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
 
-            // predicate being true means we read one record, call the callback, and then remove it
-            while (next != null && predicate.get()) {
-                if (next.getKey().time() != minTimestamp) {
```
Is this check and throwing IllegalStateException completely removed?  Was this validation unnecessary, or does removing it risk data inconsistency? If `minTimestamp` could become out of sync with the actual minimum in sortedMap, this was a useful safeguard.

aliehsaeedii commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueChangeBuffer.java:
```diff
@@ -381,55 +383,46 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
+            final Map.Entry<BufferKey, BufferValue> next = iterator.next();
+            final BufferValue bufferValue = next.getValue();
+
+            // Collect evictions to process them outside the iterator loop
```
Consider adding a comment explaining why callbacks are deferred.

aliehsaeedii replied:
// Collect evictions to process them outside the iterator loop
  // This prevents callback side effects from interfering with iteration

aliehsaeedii reviewed (commented):
Thanks @abhi-ksolves. You addressed the issue correctly I think. Now we have safer callback execution with the cost of higher memory usage. I left some minor comments. Overall LGTM.

For memory optimization, I can suggest keeping the serialized data in the eviction list. It's up to you to either apply it or not. Something like:
```
final Iterator<Map.Entry<BufferKey, BufferValue>> iterator = sortedMap.entrySet().iterator();
final List<PendingEviction> pendingEvictions = new ArrayList<>();

 private static class PendingEviction {
      final Bytes serializedKey;
      final byte[] oldValue;
      final byte[] newValue;
      final ProcessorRecordContext context;

      PendingEviction(Bytes serializedKey, BufferValue bufferValue) {
          this.serializedKey = serializedKey;
          this.oldValue = bufferValue.oldValue();
          this.newValue = bufferValue.newValue();
          this.context = bufferValue.context();
      }
  }

aliehsaeedii commented:
@abhi-ksolves could you please rebase and also address the nits? We are close to merge:)
